stdlib/sys/file.gr

```diff
@@ -1664,3 +1710,250 @@ provide let pathRename =
+            Memory.copy(computedPath + 8n, computed, computedLen)
+            WasmI32.toGrain(computedPath): String
+          }
+          Ok((fd, computedPath))
+        },
+        None => Err(Wasi.SystemError(tagSimpleNumber(Wasi._ENOENT))),
+      }
+    },
+  }
+}
+
+@unsafe
+let findPath = (path: String) => {
+  let pathPtr = WasmI32.fromGrain(path)
+  let mut pathLen = WasmI32.load(pathPtr, 4n)
+  let mut pathPtr = pathPtr + 8n
+
+  // ignore leading `/` characters
+  while (WasmI32.load8U(pathPtr, 0n) == _CHAR_SLASH) {
+    pathPtr += 1n
+    pathLen -= 1n
+  }
+
+  let preopens = match (preopens) {
+    Some(preopens) => preopens,
+    None => populatePreopens(),
+  }
+
+  let matchingPath = findPath(pathPtr, pathLen, preopens, None)
+  ignore(path)
+  matchingPath
+}
+
+// explicit just in case we ever want to support changing directories
+let cwd = "/"
+
+@unsafe
+let makeAbsolute = (path: String) => {
+  let origPathPtr = WasmI32.fromGrain(path)
+  let mut pathPtr = origPathPtr
+  let mut pathLen = WasmI32.load(pathPtr, 4n)
+  pathPtr += 8n
+
+  // already absolute
+  if (WasmI32.load8U(pathPtr, 0n) == _CHAR_SLASH) {
+    return path
+  }
+
+  // pointing at the current dirextory
+  if (
+    pathLen == 0n ||
+    pathLen == 1n && WasmI32.load8U(pathPtr, 0n) == _CHAR_DOT ||
+    pathLen == 2n &&
+    WasmI32.load8U(pathPtr, 0n) == _CHAR_DOT &&
+    WasmI32.load8U(pathPtr, 1n) == _CHAR_SLASH
+  ) {
+    return cwd
+  }
+
+  // strip leading ./
+  if (
+    WasmI32.load8U(pathPtr, 0n) == _CHAR_DOT &&
+    WasmI32.load8U(pathPtr, 1n) == _CHAR_SLASH
+  ) {
+    pathPtr += 2n
+    pathLen -= 2n
+  }
+
+  let mut cwdPtr = WasmI32.fromGrain(cwd)
+  let cwdLen = WasmI32.load(cwdPtr, 4n)
+  cwdPtr += 8n
+
+  let needSlash = WasmI32.load8U(cwdPtr + cwdLen - 1n, 0n) != _CHAR_SLASH
+
+  if (needSlash) {
+    let fullPath = allocateString(cwdLen + 1n + pathLen)
+    Memory.copy(fullPath + 8n, cwdPtr, cwdLen)
+    WasmI32.store8(fullPath + 8n + cwdLen, _CHAR_SLASH, 0n)
+    Memory.copy(fullPath + 8n + cwdLen + 1n, pathPtr, pathLen)
+    return WasmI32.toGrain(fullPath): String
+  } else {
+    let fullPath = allocateString(cwdLen + pathLen)
+    Memory.copy(fullPath + 8n, cwdPtr, cwdLen)
+    Memory.copy(fullPath + 8n + cwdLen, pathPtr, pathLen)
+    return WasmI32.toGrain(fullPath): String
+  }
+}
+
+/**
+ * Similar to `pathOpen`, but resolves the path relative to preopened directories.
+ *
+ * @param path: The path to the file or directory
+ * @param openFlags: Flags that decide how the path will be opened
+ * @param rights: The rights that dictate what may be done with the returned file descriptor
+ * @param rightsInheriting: The rights that dictate what may be done with file descriptors derived from this file descriptor
+ * @param flags: Flags which affect read/write operations on this file descriptor
+ * @returns `Ok(fd)` of the opened file or directory if successful or `Err(exception)` otherwise
```

Comment:
```suggestion
 * @returns `Ok(fd)` of the opened file or directory if successful or `Err(exception)` otherwise
 *
 * @since v0.6.0
```
This also needs throws docs because it uses populatePreopens which can throw.